### PR TITLE
refactor: refactor storage API

### PR DESCRIPTION
**Pull request by nsklikas**
IAM-1184

Refactor according to https://github.com/ory/hydra/pull/3851#issuecomment-2462440717

Do not merge, I will apply these changes directly to canonical-master when approved

### CODE REVIEW

File: storage/memory.go [hunk 1]
```diff
@@ -515,41 +515,25 @@ func (s *MemoryStore) DeletePARSession(ctx context.Context, requestURI string) (
...
-	defer s.deviceCodesMutex.Unlock()
-
-	// Only update if exist
-	if _, exists := s.DeviceCodes[signature]; exists {
-		s.DeviceCodes[signature] = req
-		s.DeviceCodesRequestIDs[req.GetID()] = signature
-	}
+	s.DeviceAuths[deviceCodeSignature] = req
+	s.DeviceAuths[userCodeSignature] = req
+	s.DeviceCodesRequestIDs[req.GetID()] = DeviceAuthPair{deviceCodeSignature, userCodeSignature}
```
Comment (shipperizer) [inline storage/memory.go]: might be wrong but my linter used to glow when using assignments like this without specifying the name of the attribute

File: storage/memory.go [hunk 2]
```diff
@@ -593,11 +565,11 @@ func (s *MemoryStore) GetUserCodeSession(_ context.Context, signature string, _
 
 // GetUserCodeSession invalidates the user code session
```
Comment (wood-push-melon) [inline storage/memory.go]: The comment needs to be updated to `InvalidateUserCodeSession`.

File: handler/rfc8628/storage.go [hunk 3]
```diff
@@ -12,16 +12,15 @@ import (
...
+	DeviceAuthStorage
 	oauth2.AccessTokenStorage
 	oauth2.RefreshTokenStorage
 }
 
-// DeviceCodeStorage handles the device_code storage
-type DeviceCodeStorage interface {
-	// CreateDeviceCodeSession stores the device request for a given device code.
-	CreateDeviceCodeSession(ctx context.Context, signature string, request fosite.Requester) (err error)
+// DeviceAuthStorage handles the device_code storage
```
Comment (wood-push-melon) [inline handler/rfc8628/storage.go]: Change to `handles the device auth session storage` or `device_code & user_code storage`?

File: handler/rfc8628/storage.go [hunk 4]
```diff
@@ -12,16 +12,15 @@ import (
...
+type DeviceAuthStorage interface {
+	// CreateDeviceAuthSession stores the device request for a given device code.
```
Comment (wood-push-melon) [inline handler/rfc8628/storage.go]: Seems this one needs to be updated too.

File: handler/rfc8628/storage.go [hunk 5]
```diff
@@ -30,24 +29,18 @@ type DeviceCodeStorage interface {
...
-	CreateUserCodeSession(ctx context.Context, signature string, request fosite.Requester) (err error)
-
 	// GetUserCodeSession hydrates the session based on the given user code and returns the device request.
 	// If the user code has been invalidated with `InvalidateUserCodeSession`, this
 	// method should return the ErrInvalidatedUserCode error.
 	//
 	// Make sure to also return the fosite.Requester value when returning the fosite.ErrInvalidatedUserCode error!
 	GetUserCodeSession(ctx context.Context, signature string, session fosite.Session) (request fosite.Requester, err error)
 
+	// InvalidateDeviceCodeSession is called when a device code is being used. The state of the user
```
Comment (wood-push-melon) [inline handler/rfc8628/storage.go]: `The state of the device code`?

Comment (nsklikas) [reply handler/rfc8628/storage.go]: thanks, for all the fixes

**Pull request closed without merging**